designdocs/recovery_system.md

```diff
@@ -0,0 +1,609 @@
+
+### Add Recovery Commitment
+
+This extrinsic allows a provider to add a new Recovery Commitment to an existing MSA. It can be used during account creation or later to update the Recovery Commitment.
+
+```rust
+// Pseudo-code for adding a Recovery Commitment to an existing MSA
+pub fn add_recovery_commitment(
+    origin: OriginFor<T>,
+    msa_owner_key: T::AccountId,
+    proof: MultiSignature,
+    payload: RecoveryCommitmentPayload,
+) -> DispatchResult {
+    let provider_key = ensure_signed(origin)?;
+
+    // Verify that the MsaId owner has signed the payload
+    ensure!(
+        Self::verify_signature(&proof, &msa_owner_key, &payload),
+        Error::<T>::InvalidSignature
+    )
+
+    // Register the signature to prevent replay attacks
+    Self::register_signature(&proof, payload.expiration.into());
+    
+    // Recover the MSA ID from the msa_owner_key
+    let msa_id = ensure_valid_msa_key(&msa_owner_key);
+
+    // Store the new RecoveryCommitment
+    MsaIdToRecoveryCommitment::<T>::insert(msa_id, payload.recovery_commitment);
+    Self::deposit_event(Event::RecoveryCommitmentAdded {
+      who: msa_owner_key,
+      msa_id,
+    });
+
+    Ok(())
+}
+```
+
+### Verify Intermediary Hashes
+
+This extrinsic is used by the Recovery Provider to verify if a user’s Recovery Intermediary Hashes compute a match to any existing MSA.
+
+```rust
+// Pseudo-code for verifying the Recovery Intermediary Hashes
+pub fn verify_intermediary_hashes(
+    payload: IntermediaryHashesPayload,
+) -> DispatchResult {
+    let who = ensure_signed(origin)?;
+    
+    // Verify that the Recovery Provider is a registered Recovery Provider
+    ensure!(
+        T::RecoveryProviderRegistry::is_recovery_provider(&who),
+        Error::<T>::NotAuthorizedRecoveryProvider
+    );
+
+    let recovery_commitment = hash_recovery_secret(payload.intermediary_hash_a, payload.intermediary_hash_b);
+
+    // TODO: How do we find the MSA ID from the hashes?
+    // Is a double index required, or is there another method?
+    let msa_id = RecoveryCommitmentToMsaId::<T>::get(&recovery_commitment)
+        .ok_or(Error::<T>::RecoveryCommitmentNotFound)?;
+
+    // Look up the Recovery Commitment in storage
+    match MsaIdToRecoveryCommitment::<T>::get(&msa_id) {
+        Some(recovery_commitment) if recovery_commitment == recovery_commitment => {
+            Self::deposit_event(Event::RecoveryCommitmentVerified { who, msa_id });
+            Ok(())
+        },
+        None => Err(Error::<T>::RecoveryCommitmentNotFound.into()),
+    }
+}
+```
+
+### Recover Account
+
+This extrinsic is called by the Recovery Provider after verifying the user’s Recovery Intermediary Hashes. It adds a new Control Key to the MSA and invalidates the old Recovery Commitment.
+
+```rust
+// Pseudo-code for recovering an account with a new control key
+pub fn recover_account(
+    origin: OriginFor<T>,
+    new_control_key: Vec<u8>,
+    payload: IntermediaryHashesPayload,
```

Comment:
I'm tempted to say we should just do two parameters instead of a payload system as it doesn't have an external signer.